### DISCUSSION TITLE
Self-host: Storage API is not working (Docker container unhealthy)

### DISCUSSION THREAD

**Discussion by seanpham99** [supabase/supabase#34343] [Troubleshooting]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I self-host Supabase with Docker on my machine as per the instructions on [Supabase Docs](https://supabase.com/docs/guides/self-hosting/docker). After successfully composing, the containers `supabase/storage-api:v1.0.6` and `supabase/studio` are in an unhealthy status. I can create buckets in the dashboard but fail to create folders in the bucket or upload files.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Follow the instructions on the Supabase self-hosting Docker guide.
2. Run docker-compose up.
3. Check the status of the containers supabase/storage-api and supabase/studio.
4. Try creating a bucket, then attempt to create a folder in the bucket or upload a file.

## Expected behavior

The `supabase/storage-api` and `supabase/studio` containers should be in a healthy status, allowing the creation of folders and uploading of files within buckets.

## Screenshots

![image](https://github.com/supabase/supabase/assets/115443243/2860a31f-7fd2-482e-9c6b-9f7e1378c947)

## System information

- Operating System: Ubuntu 22.04.4 LTS              
- Kernel: Linux 5.15.0-107-generic
- Browser (if applies) Chrome
- Version of Node.js: v20.11.1

![image](https://github.com/supabase/supabase/assets/115443243/63b1a66b-391f-4c4c-8825-7fd2e931571c)

**Comment by encima** [comment]
Thanks for opening @RoadToDev101 
This is a bug in the self-hosted docker-compose as the storage api seems not to listen on `localhost`, so it reports as unhealthy when it is fine.
This can be confirmed with `docker logs $STORAGE_CONTAINER_ID`

We will make a PR to address this. That said, the issue of not being able to create folders cannot be reproduced, can you provide logs from the storage container, please?

**Comment by seanpham99** [comment]
Hi @encima,

This is the log I get from the storage container when render the storage page on dashboard and create a new bucket name 'design_views'

`{"level":30,"time":"2024-06-18T01:26:18.129Z","pid":1,"hostname":"a122cace4088","msg":"Server listening at http://0.0.0.0:5000"}
{"level":30,"time":"2024-06-18T01:28:10.474Z","pid":1,"hostname":"a122cace4088","reqId":"req-1","tenantId":"stub","project":"stub","reqId":"req-1","type":"request","req":{"traceId":"req-1","method":"GET","url":"/bucket","headers":{"host":"storage:5000","x_forwarded_proto":"http","x_forwarded_host":"kong","x_forwarded_port":"8000","x_real_ip":"192.168.48.11","x_client_info":"supabase-js-node/2.41.1","accept":"*/*","user_agent":"node"},"hostname":"storage:5000","remoteAddress":"192.168.48.12","remotePort":43888},"res":{"statusCode":200,"headers":{"content_type":"application/json; charset=utf-8","content_length":"2"}},"responseTime":40.07513475418091,"msg":"stub | GET | 200 | 192.168.48.12 | req-1 | /bucket | node"}
{"level":30,"time":"2024-06-18T01:28:26.236Z","pid":1,"hostname":"a122cace4088","reqId":"req-2","tenantId":"stub","project":"stub","reqId":"req-2","type":"request","req":{"traceId":"req-2","method":"POST","url":"/bucket","headers":{"host":"storage:5000","x_forwarded_proto":"http","x_forwarded_host":"kong","x_forwarded_port":"8000","x_real_ip":"192.168.48.11","content_length":"107","content_type":"application/json","x_client_info":"supabase-js-node/2.41.1","accept":"*/*","user_agent":"node"},"hostname":"storage:5000","remoteAddress":"192.168.48.12","remotePort":53754},"res":{"statusCode":200,"headers":{"content_type":"application/json; charset=utf-8","content_length":"23"}},"responseTime":131.50732326507568,"msg":"stub | POST | 200 | 192.168.48.12 | req-2 | /bucket | node"}
{"level":30,"time":"2024-06-18T01:28:26.252Z","pid":1,"hostname":"a122cace4088","reqId":"req-3","tenantId":"stub","project":"stub","reqId":"req-3","type":"request","req":{"traceId":"req-3","method":"GET","url":"/bucket","headers":{"host":"storage:5000","x_forwarded_proto":"http","x_forwarded_host":"kong","x_forwarded_port":"8000","x_real_ip":"192.168.48.11","x_client_info":"supabase-js-node/2.41.1","accept":"*/*","user_agent":"node"},"hostname":"storage:5000","remoteAddress":"192.168.48.12","remotePort":53754},"res":{"statusCode":200,"headers":{"content_type":"application/json; charset=utf-8","content_length":"200"}},"responseTime":5.60968017578125,"msg":"stub | GET | 200 | 192.168.48.12 | req-3 | /bucket | node"}
{"level":30,"time":"2024-06-18T01:28:26.349Z","pid":1,"hostname":"a122cace4088","reqId":"req-4","tenantId":"stub","project":"stub","reqId":"req-4","type":"request","req":{"traceId":"req-4","method":"POST","url":"/object/list/design_views","headers":{"host":"storage:5000","x_forwarded_proto":"http","x_forwarded_host":"kong","x_forwarded_port":"8000","x_real_ip":"192.168.48.11","content_length":"89","content_type":"application/json","x_client_info":"supabase-js-node/2.41.1","accept":"*/*","user_agent":"node"},"hostname":"storage:5000","remoteAddress":"192.168.48.12","remotePort":53754},"res":{"statusCode":200,"headers":{"content_type":"application/json; charset=utf-8","content_length":"2"}},"responseTime":7.82005500793457,"msg":"stub | POST | 200 | 192.168.48.12 | req-4 | /object/list/design_views | node"}`

After that, I try to create a new folder in this bucket, the frontend shows that it created but when I reload, the folder is not there (I think it is not actually create the folder). And when I try uploading the file, it just spinning forever. And in these 2 actions, there is no log in the `storage` container.

![image](https://github.com/supabase/supabase/assets/115443243/163f61b8-d071-484e-8d2e-ed82c3b07fdf)

Plus the studio container also unhealthy, but when check the log, it seems there is no problem. I hope this help.

![image](https://github.com/supabase/supabase/assets/115443243/f940b3b5-e019-4728-aa1b-8c840f71358d)

**Comment by encima** [comment]
thanks @RoadToDev101 , those logs show success for the storage requests so that certainly is weird! 
Is there some issue with the volume mounting? Are you able to create a folder when you call the storage API directly?

**Comment by seanpham99** [comment]
Hi @encima,

I try to fetch the bucket api on Postman but I cant due to "Invalid authentication credentials". I try both Anon and Service Key for authorization. See the image below

![image](https://github.com/supabase/supabase/assets/115443243/600e673d-381c-473b-abd3-973d997e6d02)

The jwt in .env :
`JWT_SECRET=girSzUKW1c937xjnJhKhuqLLM4WPIjNXrV8Odvra
ANON_KEY=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.ewogICJyb2xlIjogImFub24iLAogICJpc3MiOiAic3VwYWJhc2UiLAogICJpYXQiOiAxNzE4NjQzNjAwLAogICJleHAiOiAxODc2NDEwMDAwCn0.STjbgWo03Cn8SRaRAn2bqw0M1io-k6g8GJkHAeUwXY4
SERVICE_ROLE_KEY=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.ewogICJyb2xlIjogInNlcnZpY2Vfcm9sZSIsCiAgImlzcyI6ICJzdXBhYmFzZSIsCiAgImlhdCI6IDE3MTg2NDM2MDAsCiAgImV4cCI6IDE4NzY0MTAwMDAKfQ.aq4Nv_JAsUJAJxwvr5Ro6GDyXVbyoGb9ij2UN40kJ_Y`

Am I do it wrong?

![image](https://github.com/supabase/supabase/assets/115443243/724c4cbb-d1e9-4a75-82ab-1d5509680be3)

**Comment by seanpham99** [comment]
@encima, so strange... I can upload the image through my NodeJS app.

![image](https://github.com/supabase/supabase/assets/115443243/bf9e8619-9f62-4e56-aaa5-ffe8c55ae797)

**Comment by encima** [comment]
@RoadToDev101 thanks for the update. so it works using the supabase client is that right? does it still fail with the dashboard? 

from Postman, it looks like the credentials are probably failing due to them being incorrectly encoded or not picked up properly.

**Comment by bachkukkik** [comment]
for `supabase/studio`, I think it is because of the starting period

I went inside the docker container and command

`$ node -e "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"`

it gave me no error so i increased timeout in health check condition as below to fix unhealthy

```
supabase-studio:
  image: 'supabase/studio:20240422-5cf8f30'
  healthcheck:
    test:
      - CMD
      - node
      - '-e'
      - "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"
    timeout: 15s
    interval: 5s
    retries: 5
  depends_on:
    supabase-analytics:
      condition: service_healthy
```

However it doesn't work with storage-api. I tried to debug the same way but got below result
```
/app # wget --no-verbose --tries=1 --spider http://localhost:5000/status
Connecting to localhost:5000 ([::1]:5000)
wget: can't connect to remote host: Connection refused
```

**Comment by encima** [comment]
> for `supabase/studio`, I think it is because of the starting period
> 
> I went inside the docker container and command
> 
> `$ node -e "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"`
> 
> it gave me no error so i increased timeout in health check condition as below to fix unhealthy
> 
> ```
> supabase-studio:
>   image: 'supabase/studio:20240422-5cf8f30'
>   healthcheck:
>     test:
>       - CMD
>       - node
>       - '-e'
>       - "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"
>     timeout: 15s
>     interval: 5s
>     retries: 5
>   depends_on:
>     supabase-analytics:
>       condition: service_healthy
> ```
> 
> However it doesn't work with storage-api. I tried to debug the same way but got below result
> 
> ```
> /app # wget --no-verbose --tries=1 --spider http://localhost:5000/status
> Connecting to localhost:5000 ([::1]:5000)
> wget: can't connect to remote host: Connection refused
> ```

Are you connecting to storage from within the studio container? Either way, `localhost` will not work as it is not exposed there (unless on the storage container). You can use the same command but use `studio` instead of `localhost`

**Comment by seanpham99** [comment]
@encima I use the healthcheck command in the docker compose file for `studio` container and get this error:
![image](https://github.com/supabase/supabase/assets/115443243/2448ef15-0e5e-4a70-9e77-cb44817a24ac)

**Comment by encima** [comment]
> @encima I use the healthcheck command in the docker compose file for `studio` container and get this error:
> 
> ![image](https://github.com/supabase/supabase/assets/115443243/2448ef15-0e5e-4a70-9e77-cb44817a24ac)
> 
> 

@RoadToDev101 that is expected yes. the studio is accessible through "studio" but not localhost.

**Comment by major-winter** [comment]
I'm also facing the unhealthy issues on those two services. In addition, I also got the unauthorized issue when uploading files that are more than 6 MB; though, it works fine for those that are less than that.

**Comment by BierDav** [comment]
Same issue here, the nginx server doesn't serve on `localhost` which makes the healthcheck provided in the github repo not working. Is there a workaround? Another healthcheck command. I tried to comeup with one, but without the ability to fetch from localhost it is kinda difficult to not hard code an IP.

**Comment by bachkukkik** [comment]
> for `supabase/studio`, I think it is because of the starting period
> 
> I went inside the docker container and command
> 
> `$ node -e "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"`
> 
> it gave me no error so i increased timeout in health check condition as below to fix unhealthy
> 
> ```
> supabase-studio:
>   image: 'supabase/studio:20240422-5cf8f30'
>   healthcheck:
>     test:
>       - CMD
>       - node
>       - '-e'
>       - "require('http').get('http://localhost:3000/api/profile', (r) => {if (r.statusCode !== 200) throw new Error(r.statusCode)})"
>     timeout: 15s
>     interval: 5s
>     retries: 5
>   depends_on:
>     supabase-analytics:
>       condition: service_healthy
> ```
> 
> However it doesn't work with storage-api. I tried to debug the same way but got below result
> 
> ```
> /app # wget --no-verbose --tries=1 --spider http://localhost:5000/status
> Connecting to localhost:5000 ([::1]:5000)
> wget: can't connect to remote host: Connection refused
> ```


Just to revisit my previous attempt, as of now i manage to get both studio and storage services a status of healthy. Note that if tag `latest` doesn't work, you may need to specify a tag for example `supabase/studio:20250113-83c9420`

```yaml
  supabase-studio:
    image: 'supabase/studio:20250113-83c9420'
    healthcheck:
      test:
        [
          "CMD",
          "node",
          "-e",
          "fetch('http://localhost:3000/api/profile').then((r) => {if (r.status !== 200) throw new Error(r.status)})"
        ]
      timeout: 15s
      interval: 5s
      retries: 5

  supabase-storage:
    image: 'supabase/storage-api:latest'
    healthcheck:
      test:
        [
          "CMD",
          "node",
          "-e",
          "fetch('http://localhost:5000/status').then((r) => {if (r.status !== 200) throw new Error(r.status)})"
        ]
      timeout: 15s
      interval: 5s
      retries: 5
```

**Comment by nthtrung09it** [comment]
tks @bachkukkik, I use that version **supabase/studio:20250113-83c9420** and it works.

**Comment by Hallidayo** [comment]
Hi everyone, due to inactivity on this issue I've moved the issue over to discussions for troubleshooting for other users, Thankyou.
